Include openaddresses data by pattern

remi-dupre opened the pull request:
Change the download logic for OpenAddresses: instead of downloading datasets atomically, we download the full released file from OA and filter content using more generic patterns.

**EDIT**

I've updated this PR to only re-download files periodically (or when the checksum changed in the case of OSM).

Also, I removed `data_src` setting from the settings of the image `download`. Instead, calls to the downloader must provide parameters with correct paths to the volumes.

amatissart commented on download/tasks.py:
```diff
@@ -20,25 +21,36 @@ def download_bano(ctx, bano_url, output_file):
+    for dirname, _, files in walk(oa_tmp_dir):
+        for filename in files:
+            full_path = path.join(dirname, filename)
+            relt_path = full_path.lstrip(oa_tmp_dir)
```
`.lstrip()` takes a set of chars as its argument, so is actually not safe in this context.
```suggestion
            relt_path = os.path.relpath(full_path, oa_tmp_dir)
```

amatissart commented on download/tasks.py:
```diff
@@ -20,25 +21,36 @@ def download_bano(ctx, bano_url, output_file):
-        flat_csv_filename = csv_file.lstrip(oa_temp_dir).replace("/", "__")
-        ctx.run(f"mv {csv_file} {output_dir}/{flat_csv_filename}")
+    for filename in included_files:
+        flat_name = filename.lstrip(oa_tmp_dir).replace("/", "__")
```
the same here

remi-dupre commented:
I've updated this PR to only re-download files periodically (or when the checksum changed in the case of OSM).

Also, I removed `data_src` setting from the settings of the image `download`. Instead calls to the downloader must provide parameters with correct paths to the volumes.

remi-dupre commented:
I've updated this PR to only re-download files periodically (or when the checksum changed in the case of OSM).

Also, I removed `data_src` setting from the settings of the image `download`. Instead calls to the downloader must provide parameters with correct paths to the volumes.

GuillaumeGomez commented on download/tasks.py:
```diff
@@ -1,44 +1,151 @@
+
+def needs_to_download(ctx, filename, max_age=None, md5_url=None):
+    """
+    Check if "filename" should be downloaded because it doesn't exist, or is
```
'"it doesn't exist, or is" -> '"it doesn't exist or is"

amatissart commented on download/invoke.yml:
```diff
@@ -1,3 +1,12 @@
+tmp_dir: /tmp
+
+# URL for the zip archive containing all data.
+oa_url: https://data.openaddresses.io/openaddr-collected-europe.zip
```
This URL does not seem to be configurable from the top-level invoke configuration.  
For example, we could want to download the planet file instead.

remi-dupre replied:
I moved this piece of configuration in the top-level `invoke.yml`, it seems indeed cleaner to have everything in the same place :)

amatissart commented on download/tasks.py:
```diff
@@ -1,44 +1,151 @@
+        return True
+
+    if md5_url is not None:
+        res = requests.get(md5_url)
```
HTTP errors should be caught here.  
For example, the file should be downloaded if the md5 is not found.

amatissart commented on download/tasks.py:
```diff
@@ -1,44 +1,151 @@
+        md5 = hashlib.md5()
+        for chunk in iter(lambda: f.read(4096), b""):
+            md5.update(chunk)
+        md5 = md5.hexdigest()
```
Maybe we should check the integrity of the downloaded file, thanks to this md5 ?

remi-dupre replied:
It will now stop the execution when this happens, this is a bit brutal but at least we are sure not to miss the issue ?

Maybe we could try to download a few more times before crashing ?

amatissart replied:
Hum yes, that would be nice. But this case should be very rare.  
Let's keep that in mind for another issue, I don't want to delay this PR for too long.
